### ISSUE TITLE
Chaining Html.App.map messes up

### ISSUE THREAD

**Issue by evancz** [elm-lang/virtual-dom#21]
As reported by @pdamoc in https://github.com/elm-lang/html/issues/16 and further clarified by @debois.

```elm
import Html exposing (Html, button, div, text)
import Html.App as App
import Html.Events exposing (onClick)


main : Program Never
main =
  App.beginnerProgram
    { model = 0
    , update = always <| (+) 1
    , view = view
    }


view : Int -> Html ()
view model =
  div
    []
    [ text (toString model)
    , button [ onClick () ] [ text "Inc" ]
    ]
  |> App.map identity
  |> App.map identity
```

Notes. 

  1. The repeated `App.map` is necessary to trigger the error. 
  2. Adding a second button before the `text` (see below) suppresses the JS error, but causes updates to change the text on the left-most button instead of the `text` element:

```elm
  [ button [ onClick () ] [ text "Inc" ] 
  , text (toString model)
  , button [ onClick () ] [ text "Inc" ]
  ]
```

**Comment by vito** [comment]
I'm also seeing interesting behavior with nested components, which I thought would be fixed by https://github.com/elm-lang/core/commit/cce1dc54f15ff9bae5d2900faf2661f2d1cd83b2 - but it wasn't. With a component nesting another component and doing `Html.App.map` to wrap all actions coming from the sub-component, the outer `update` is being called with the sub-component's messages *not* wrapped, which should be impossible.

The pattern-matching then does unpredictable things, as it often goes down the `default:` branch, which could do all sorts of bad things since it's assuming that `default:` branch matched the only remaining pattern (common example: accessing fields that don't exist, and assigning them on the model as `undefined`).

I'm seeing this in a pretty [complicated stack](https://github.com/concourse/atc/tree/54bcb2eef07e8d794de71d7555d2c34c2b22a1a6) - so I'll try to whittle it down to a smaller repro. Just wanted to follow-up to see if any of this sounds related.

**Comment by pdamoc** [comment]
@vito,  is your problem similar to my the second context I described in  elm-lang/html/#16 ?
Here is a simplified version:
https://gist.github.com/pdamoc/27e4198940d50e0e38b5fba05304d799
Build WidgetList.elm and then: type "cat", hit enter, type "dog", hit enter, type anything.

**Comment by blakesweeney** [comment]
I've also run into the same behavior as @vito in a fairly complicated stack on 0.17. I'll try to create a small example.

**Comment by jvoigtlaender** [comment]
Fixed by https://github.com/elm-lang/virtual-dom/pull/20?

**Comment by evancz** [comment]
Should be fixed by #20
